App/Services/RpcController.cs

```diff
@@ -0,0 +1,248 @@
+                    ApiToken = credentials.ApiToken,
+                },
+            }, ct);
+            if (reply.MsgCase != ServiceMessage.MsgOneofCase.Start)
+                throw new InvalidOperationException($"Unexpected reply message type: {reply.MsgCase}");
+        }
+        catch (Exception e)
+        {
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+            throw new RpcOperationException("Failed to send start command to service", e);
+        }
+
+        if (!reply.Start.Success)
+        {
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+            throw new VpnLifecycleException("Failed to start VPN",
+                new InvalidOperationException($"Service reported failure: {reply.Start.ErrorMessage}"));
+        }
+
+        MutateState(state => { state.VpnLifecycle = VpnLifecycle.Started; });
+    }
+
+    public async Task StopVpn(CancellationToken ct = default)
+    {
+        using var _ = await AcquireOperationLockNowAsync();
+        EnsureRpcConnected();
+
+        MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopping; });
+
+        ServiceMessage reply;
+        try
+        {
+            reply = await _speaker!.SendRequestAwaitReply(new ClientMessage
+            {
+                Stop = new StopRequest(),
+            }, ct);
+        }
+        catch (Exception e)
+        {
+            throw new RpcOperationException("Failed to send stop command to service", e);
+        }
+        finally
+        {
+            // Technically the state is unknown now.
+            MutateState(state => { state.VpnLifecycle = VpnLifecycle.Stopped; });
+        }
+
+        if (reply.MsgCase != ServiceMessage.MsgOneofCase.Stop)
+            throw new VpnLifecycleException("Failed to stop VPN",
+                new InvalidOperationException($"Unexpected reply message type: {reply.MsgCase}"));
+        if (!reply.Stop.Success)
+            throw new VpnLifecycleException("Failed to stop VPN",
+                new InvalidOperationException($"Service reported failure: {reply.Stop.ErrorMessage}"));
+    }
+
+    private void MutateState(Action<RpcModel> mutator)
+    {
+        using var _ = _stateLock.Lock();
+        mutator(_state);
+        StateChanged?.Invoke(this, _state.Clone());
+    }
+
+    private async Task<IDisposable> AcquireOperationLockNowAsync()
+    {
+        var locker = await _operationLock.LockAsync(TimeSpan.Zero);
+        if (locker == null)
+            throw new InvalidOperationException("Cannot perform operation while another operation is in progress");
+        return locker;
+    }
+
+    private void SpeakerOnReceive(ReplyableRpcMessage<ClientMessage, ServiceMessage> message)
+    {
+        // TODO: this
+    }
+
+    private void SpeakerOnError(Exception e)
+    {
+        Debug.WriteLine($"Error: {e}");
+        Reconnect(CancellationToken.None).Wait();
+    }
+
+    private void EnsureRpcConnected()
```

Comment:
To me, "ensure" reads as, check if the thing is how we need it, and if not, fix it. As in, "ensure you bring your umbrella." This just checks and throws up if it's wrong, so maybe "assert"?